optimize-onnx/src/optimize.cpp

```diff
@@ -0,0 +1,176 @@
+// or the predict net. If you want both, which you probably do,
+// arrange to call it twice.
+//
+// NOTE POTENTIAL BREAKAGE:
+//
+// The ONNX spec provides no guarantees about "staging", i.e. which
+// inputs change on every invocation vs which generally stay the same.
+// Here we make the assumption that inputs which have an initializer
+// value provided for them vary only between invocations of the init
+// net, and are constant across runs of the predict net.
+//
+std::shared_ptr<Graph> split_init_and_predict(std::shared_ptr<Graph> g, bool init, bool predict) {
+  // The first step is a graph traversal to identify what is reachable
+  // from the mutable intializers. Since we have a bipartite graph, we
+  // must track both Nodes and Values that we have reached (as either
+  // could be terminal objects in the graph), but it's sufficient to
+  // maintain a frontier consisting only of Values.
+  std::unordered_set<Value *> live_values;
+  std::unordered_set<Node *> live_nodes;
+
+  {
+    std::queue<Value *> frontier;
+
+    {
+      std::unordered_set<std::string> initializer_names(
+        g->initializer_names().begin(),
+        g->initializer_names().end());
+
+      // Seed frontier with graph input Values that do not have a
+      // corresponding Initializer.
+      for (Value * v : g->inputs()) {
+        if (initializer_names.find(v->uniqueName()) == initializer_names.end()) {
+          live_values.insert(v);
+          frontier.push(v);
+        }
+      }
+    }
+
+    // Additionally seed frontier with the output Values of impure
+    // Nodes.
+    for (Node * n : g->nodes()) {
+      if (isPureOperator(n)) {
+        continue;
+      }
+      live_nodes.insert(n);
+      for (Value * v : n->outputs()) {
+        live_values.insert(v);
+        frontier.push(v);
+      }
+    }
+
+    // Mark as live all Values and Nodes that are reachable from the
+    // seed Values and Nodes.
+    while (!frontier.empty()) {
+      for (Use u : frontier.front()->uses()) {
+        if (live_nodes.find(u.user) != live_nodes.end()) {
+          continue;
+        }
+        live_nodes.insert(u.user);
+        for (Value * v : u.user->outputs()) {
+          if (live_values.find(v) != live_values.end()) {
+            continue;
+          }
+          live_values.insert(v);
+          frontier.push(v);
+        }
+      }
+      frontier.pop();
+    }
+  }
+
+  if (init) {
+    // Any non-live inputs to live nodes will become an output of the
+    // init graph.
+    std::unordered_set<Value *> new_outputs;
+    for (Node * n : live_nodes) {
+      for (Value * v : n->inputs()) {
+        if (live_values.find(v) == live_values.end()) {
+          new_outputs.insert(v);
+        }
+      }
+    }
+
+    // remove all live outputs
+    for (int i = g->outputs().size() - 1; i >= 0; i--) {
+      if (live_values.find(g->outputs()[i]) != live_values.end()) {
+        g->return_node()->eraseOutput(0);
```

Comment:
I think this should be `eraseInput(0)`, the return_node should have 0 outputs. It would be a good idea to just add  `graph.eraseOutput(0)` as a method on graph since it is confusing that outputs are represented as inputs to the return_node.